lib/SILOptimizer/Mandatory/Differentiation.cpp

```diff
@@ -1612,16 +1614,28 @@ reapplyFunctionConversion(SILValue newFunc, SILValue oldFunc,
     for (auto arg : pai->getArguments())
       newArgs.push_back(substituteOperand(arg));
     auto innerNewFunc = reapplyFunctionConversion(
-        newFunc, oldFunc, pai->getCallee(), builder, loc, substituteOperand);
-    return builder.createPartialApply(
-        loc, innerNewFunc, pai->getSubstitutionMap(), newArgs,
+        newFunc, oldFunc, pai->getCallee(), builder, loc, substMap,
+        substituteOperand);
+    if (substMap.empty()) {
+      substMap = pai->getSubstitutionMap();
+    } else {
+      substMap = substMap.subst(
```

Comment:
This doesn't look right. Do the two substitution map's generic signatures only differ in their conformance requirements?

Reply:
`substMap.subst(pai->getSubstitutionMap())` doesn't quite work, actually.
Substitution map logic is a bit convoluted because a derivative function's substitution map can have more requirements than its original function's.

Derivative functions's substitution map:
```
(substitution_map generic_signature=<τ_0_0 where τ_0_0 : Differentiable, τ_0_0 : FloatingPoint, τ_0_0 : TensorFlowScalar>
  (substitution τ_0_0 -> τ_0_0)
  (conformance type=τ_0_0
    (abstract_conformance protocol=Differentiable))
  (conformance type=τ_0_0
    (abstract_conformance protocol=FloatingPoint))
  (conformance type=τ_0_0
    (abstract_conformance protocol=TensorFlowScalar)))
```

Original function's substitution map:
```
(substitution_map generic_signature=<T where T : FloatingPoint, T : TensorFlowScalar>
  (substitution T -> Float)
  (conformance type=T
    (normal_conformance type=Float protocol=FloatingPoint lazy))
  (conformance type=T
    (normal_conformance type=Float protocol=TensorFlowScalar lazy)))
```

Goal substitution map (this is what's generated by current code, preserves requirements from the derivative):
```
(substitution_map generic_signature=<τ_0_0 where τ_0_0 : Differentiable, τ_0_0 : FloatingPoint, τ_0_0 : TensorFlowScalar>
  (substitution τ_0_0 -> Float)
  (conformance type=τ_0_0
    (normal_conformance type=Float protocol=Differentiable lazy))
  (conformance type=τ_0_0
    (normal_conformance type=Float protocol=FloatingPoint lazy))
  (conformance type=τ_0_0
    (normal_conformance type=Float protocol=TensorFlowScalar lazy)))
```

`substMap.subst(pai->getSubstitutionMap())` (crashes with error below)
```
Invalid conformance substitution
UNREACHABLE executed at /Users/danielzheng/swift-dev/swift/lib/AST/ProtocolConformance.cpp:131
swift::ProtocolConformanceRef::subst(swift::Type, llvm::function_ref<swift::Type (swift::SubstitutableType*)>, llvm::function_ref<llvm::Optional<swift::ProtocolConformanceRef> (swift::CanType, swift::Type, swift::ProtocolDecl*)>) const + 612
7  swift                    0x0000000103dc36e4 swift::ProtocolConformanceRef::subst(swift::Type, llvm::function_ref<swift::Type (swift::SubstitutableType*)>, llvm::function_ref<llvm::Optional<swift::ProtocolConformanceRef> (swift::CanType, swift::Type, swift::ProtocolDecl*)>) const + 612
8  swift                    0x0000000103dd0f37 swift::SubstitutionMap::subst(llvm::function_ref<swift::Type (swift::SubstitutableType*)>, llvm::function_ref<llvm::Optional<swift::ProtocolConformanceRef> (swift::CanType, swift::Type, swift::ProtocolDecl*)>) const + 823
9  swift                    0x0000000103dd10ab swift::SubstitutionMap::subst(swift::SubstitutionMap) const + 43
10 swift                    0x00000001034485db reapplyFunctionConversion(swift::SILValue, swift::SILValue, swift::SILValue, swift::SILBuilder&, swift::SILLocation, swift::SubstitutionMap, std::__1::function<swift::SILValue (swift::SILValue)>) + 2539
11 swift                    0x00000001034465fa emitAssociatedFunctionReference((anonymous namespace)::ADContext&, swift::SILBuilder&, (anonymous namespace)::DifferentiationTask const*, swift::SILAutoDiffIndices, swift::AutoDiffAssociatedFunctionKind, swift::SILValue, (anonymous namespace)::DifferentiationInvoker, std::__1::function<void ((anonymous namespace)::DifferentiationTask*)>) + 1338
12 swift                    0x0000000103439784 (anonymous namespace)::Differentiation::run() + 1892
13 swift                    0x000000010352de0b swift::SILPassManager::runModulePass(unsigned int) + 1083
14 swift                    0x000000010352e8e4 swift::SILPassManager::execute() + 692
15 swift                    0x0000000102a0f35b swift::SILPassManager::executePassPipelinePlan(swift::SILPassPipelinePlan const&) + 187
16 swift                    0x0000000103537794 swift::runSILDiagnosticPasses(swift::SILModule&) + 132
17 swift                    0x00000001028baf8b performCompile(swift::CompilerInstance&, swift::CompilerInvocation&, llvm::ArrayRef<char const*>, int&, swift::FrontendObserver*, swift::UnifiedStatsReporter*) + 11291
18 swift                    0x00000001028b72dd swift::performFrontend(llvm::ArrayRef<char const*>, char const*, void*, swift::FrontendObserver*) + 3021
19 swift                    0x0000000102868eae main + 686
20 libdyld.dylib            0x00007fff78b12015 start + 1
21 libdyld.dylib            0x000000000000000a start + 2270093302
```

---

Do you have advice for improving the substitution map logic such that the goal is achievable?

Reply:
Oh I see. It looks like the substMap parameter here is always assocFn->getForwardingSubstitutionMap(), which is also why you need the GenericEnvironment. This isn't quite how the APIs are meant to to be used. Instead, try this:

```
auto substMap = SubstitutionMap::get(
    assocFn->getLoweredFunctionType()->getGenericSignature(),
    QuerySubstitutionMap(pai->getSubstitutionMap()),
    LookUpConformanceInModule(builder.getModule().getSwiftModule()))
```

This avoids several intermediate steps.

Reply:
(You'll need to plumb assocFn or assocFn's generic signature down to here)

Reply:
Thank you so much!
Changed to suggested substitution map logic in https://github.com/apple/swift/pull/22240/commits/8b68594cded4de1ede729c7b4abe4575ddea81e0.
Also reverted `assocFn` creation so that it's created with a null generic environment.

Reply:
We should probably do the same in PrimalGen::visitApplyInst

Reply:
Yes, I'll change that soon.